docs/madr/decisions/053-sni.md

```diff
@@ -0,0 +1,57 @@
+# SNI for MeshService, MeshExternalService etc.
+
+* Status: accepted
+
+## Context and Problem Statement
+
+Currently, the SNI format is `<kuma.io/service>{mesh=<mesh>,tag1=value1}`. For example:
+* `backend{mesh=demo}`
+* `backend{mesh=demo,version=2}`
+
+We need to introduce SNI format for `MeshService` and `MeshExternalService` etc.
+Current format does not work, because we need to differentiate between `kuma.io/service` and MeshService.
+Current format also has other problems:
+* It violates [SNI spec](https://datatracker.ietf.org/doc/html/rfc3546#page-8). It should be a valid hostname. Using characters like `,{/` is not valid.
+  Envoy does not complain about it right now, but it can bite us in the future.
+* It can be too long. See https://github.com/kumahq/kuma/issues/8915
+
+## Considered Options
+
+* New format
+
+## Decision Outcome
+
+Chosen option: "New format", because it solves existing problem and lets us differentiate MeshService from old service.
+
+## Pros and Cons of the Options
+
+### New format
+
+It's important to note that we never parse SNI. We only use it for multizone communication on zone proxies to associate to which service pass the data.
+It means that we can hash the data as long as it's consistent hashing.
+We cannot put tags in the original form because they can violate hostname requirements. For example `/` in `kuma.io/service` is not ok.
+The new format could look like this then:
+`<resource-name>.<mesh-name>.<extra-hash-if-needed>.<resource-type>`
+* `<resource-name>` - name of the resource. Name of MeshService or MeshExternalService
+* `<mesh-name>` - name of the mesh
+* `<extra-hash-if-needed>` - 16 characters of SHA256. In case of MeshService tags it's `SHA256(<tag1>=<value1>,<tag2>=<value2>)` sorted by keys.
```

Comment:
Should we fit a version inside to easily change this in the future (.e.g not convinced sha256 is the best hash here)?

Reply:
```
const base62Charset = "0123456789abcdefghijklmnopqrstuvwxyz"

func base62Encode(input []byte) string {
	var result []byte
	num := new(big.Int).SetBytes(input)
	base := big.NewInt(int64(len(base62Charset)))

	for num.Sign() > 0 {
		mod := new(big.Int)
		num.DivMod(num, base, mod)
		result = append([]byte{base62Charset[mod.Int64()]}, result...)
	}

	return string(result)
}

func main() {
	h := sha256.New()
	bytes := h.Sum(nil)
	x := fmt.Sprintf("%x", bytes)
	base62 := base62Encode(bytes)
	fmt.Printf("%s %d %s %d", x, len(x), base62, len(base62))
}
```

```
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 64 5oaq0bjhj6un82wg98mgigso5q7qlhc63je4gw7ivixqqhkd3p 50
```

There is a little win with this at the cost of this custom implem (there seems no to be base62 lib using which we could specify charset. I'd lean towards sticking to hexadecimal, but if you think it's worth it. We can roll this implem.

Reply:
@lobkovilya used something from k8s I think for policies hash

Reply:
I added https://github.com/kumahq/kuma/blob/4ad47f000116872bf8b25676cd5ccc3cb115eed6/pkg/util/k8s/name_converter.go#L34 to copy how pod template hash is calculated, but its only 10 characters.

Reply:
> @lobkovilya used something from k8s I think for policies hash

It's `fnv.New64a()` + encoding from kube which is 
```
	// We omit vowels from the set of available characters to reduce the chances
	// of "bad words" being formed.
	alphanums = "bcdfghjklmnpqrstvwxz2456789"
```
fnv64 produces 16 chars in hex. Then some of them are exchanged to alphanums listed above.

We could use it for tags I think instead of SHA256 + trimming.